docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1058 @@
+    increment a generation number.
+
+- **Object keys are suffixed** with the generation number
+- **Safety for multiply-attached tenants** is provided by the
+  tenant attach generation in the object key: the competing pageservers will not
+  try to write to the same keys.
+- **Safety in split brain for multiple nodes running with
+  the same node ID** is provided by the pageserver calling out to the control plane
+  on startup, to increment the generations of any attached tenants
+- **Safety for deletions** is provided by pageservers validating with control plane that neither pageserver process nor the attachment has been superseded since it last updated a timeline's index.
+- **The control plane is used to issue generation numbers** to avoid the need for
+  a built-in consensus system in the pageserver, although this could in principle
+  be changed without changing the storage format.
+
+### Generation numbers
+
+A generation number is associated with each tenant in the control plane,
+and each time the attachment status of the tenant changes, this is incremented.
+Changes in attachment status include:
+
+- Attaching the tenant to a different pageserver
+- A page server restarting, and "re-attaching" its tenants on startup
+
+These increments of attachment generation provide invariants we need to avoid
+split-brain issues in storage:
+
+- If two pageservers have the same tenant attached, they are guaranteed to have different attachment generation numbers, because the generation would increment
+  while attaching the second one.
+- If there are multiple pageservers running with the same node ID, they are guaranteed to have different generation numbers, because the generation would increment
+  when the second node started and re-attached its tenants.
+- If there are multiple pageservers running with the same node ID, we may unambiguously know which
+  of them "wins" by picking the higher generation number.
+
+**Note** If we ever run pageservers on infrastructure that might transparently restart
+a pageserver while leaving an old process running (e.g. a VM gets rescheduled
+without the old one being fenced), then there is a risk of corruption, when
+the control plane attaches the tenant, as follows:
+
+- if the control plane sends an /attach request
+  to node A, then node A dies and is replaced, and the control plane times out
+  the request and restarts, then it could end up with two physical nodes both using
+  the same attachment generation.
+- This is not an issue when using EC2 instances with ephemeral storage, as long
+  as the control plane never re-uses a node ID, but it would need re-examining
+  if running on different infrastructure.
+- To robustly protect against this class of issue, we would either:
+  - add a "node generation" to distinguish between different processes holding the
+    same node_id.
+  - or, dispense with static node_id entirely and issue an ephemeral ID to each
+    pageserver process when it starts.
```

Comment:
"issue an ephemeral ID" is too detailed here.


```suggestion
- To robustly protect against this class of issue, we would either:
  - add a "node generation" to distinguish between different processes holding the
    same node_id.
  - or, dispense with the concept of static node_id entirely, and instead change
    pageserver to be completely ephemeral. In that world, a pageserver process
    always starts up without any attachments and makes itself known to control
    plane through a best-effort correlation ID, such as the EC2 instance ID.
    To control plane, this is comparable to an explicit failover request.
    The pageserver process will re-use layer files in local NVMe that were
    downloaded by the previous process on the same machine.
```

Reply:
Okay, but I'm not aiming to define the ephemeral ID system in detail in this PR: I think it's okay to refer to it in an offhand way, and there are later sections that go into morer detail.